bdk_core/src/io.rs

```diff
@@ -0,0 +1,249 @@
+    fn from(error: UpdateFailure) -> IOError {
+        IOError::TxUpdateError(error)
+    }
+}
+
+const UNCONFIRMED_TX_HEIGHT: u32 = u32::MAX - 1;
+
+/// write SparseChain to disk for persistence
+pub fn write_sparse_chain<T: Write + Seek + Read>(chain: &SparseChain, writer: T) -> Result<(), IOError> {
+    let bytes = writer.bytes().collect::<io::Result<Vec<u8>>>()?;
+    let mut cursor = io::Cursor::new(bytes);
+    // Todo: check if the state_id of the in-memory store is different from disk store
+    // if that is the case, write new data to disk, if not return because there is
+    // nothing to write.
+
+    // update version
+    // TODO: write current db version
+    let _size = cursor.write(&[0u8; 4])?;
+
+    // update state_id
+    // TODO: write current state_id
+    let _size = cursor.write(&[0u8; 4])?;
+
+    
+    // TODO: Figure out what to do if the checkpoint is None
+    if let Some(checkpoint_limit) = chain.checkpoint_limit {
+        let _size = cursor.write(&u32::to_be_bytes(checkpoint_limit as u32))?;
+
+        let mut iter_checkpoints = chain.checkpoints.iter();
+        for _i in 0..checkpoint_limit {
+            if let Some((height, block_hash)) = iter_checkpoints.next() {
+                let mut buf = [0u8; 36];
+                u32::to_be_bytes(*height).iter().enumerate().for_each(|(i, val)| buf[i] = *val);
+                encode::serialize(block_hash).iter().enumerate().for_each(|(i, val)| buf[i + 4] = *val);
+                let _size = cursor.write(&buf)?;
+            } else {
+                let zero_pad_buf = [0u8; 36];
+                let _size = cursor.write(&zero_pad_buf)?;
+            }
+        }
+    }
+
+    for (height, txid) in &chain.txid_by_height {
+        let mut buf = match height {
+            TxHeight::Confirmed(block_height)=> u32::to_be_bytes(*block_height).to_vec(),
+            TxHeight::Unconfirmed => u32::to_be_bytes(UNCONFIRMED_TX_HEIGHT).to_vec()
+        };
+        buf.extend(encode::serialize(txid));
+        let _size = cursor.write(&buf)?;
+    }
+
+    Ok(())
+}
+
+/// Instantiate SparseChain from disk bytes.
+pub fn instantiate_sparse_chain<T: Read + Seek>(chain: &mut SparseChain, reader: &mut T) -> Result<(), IOError> {
+    let bytes = reader.bytes().collect::<io::Result<Vec<u8>>>()?;
+    let mut cursor = io::Cursor::new(bytes);
+    // skip the state_id + version
+    cursor.set_position(8);
+    let mut checkpoint_limit_bytes = [0u8; 4];
+    let _size = cursor.read(&mut checkpoint_limit_bytes)?;
+    let checkpoint_limit = u32::from_be_bytes(checkpoint_limit_bytes);
+    let mut checkpoint_bytes = Vec::with_capacity((checkpoint_limit as u32 * 36u32) as usize);
+    // read checkpoint_limit * 36 bytes
+    let _size = cursor.read(&mut checkpoint_bytes)?;
+    // create checkpoints from these bytes.
+    for chunk in checkpoint_bytes.chunks(36) {
+        // once you meet the padded zeros stop reading
+        if chunk.iter().all(|x| *x == 0u8) {
```

Comment:
you can just check against ` == [0u8;36]`?

Reply:
Right